_posts/2024-11-22-cheri-myths-safety-critical.markdown

```diff
@@ -0,0 +1,183 @@
+It is absolutely not acceptable to claim that you have a safety-critical system if that safety depends on behaviour that is not specified.
+
+In the second case, the only change is the trap reason and the fact that the trap on a CHERI system gives you more information.
+
+In the third case, the failure won't appear to be memory safety today, it will be some random data corruption.
+As I wrote above, this means that you're in the 'spooky action at a distance' world.
+There is absolutely no guarantee that the corrupted data is not even more safety critical.
+For example, an out-of-bounds store that overwrites the dose value in an insulin pump will not crash on a non-CHERI system, but might well kill the patient.
+Again, if your 'safety' guarantees are framed as 'something bad happens and we don't know what, but it hasn't killed anyone yet' then you do not have safety.
+
+In the fourth case, CHERI improves things because now it should show up in testing reliably and you can fix the issue.
+
+The baseline is not better in *any* of these cases.
+The example that people often use here is anti-lock braking systems (ABS).
+You absolutely do not want the controller for ABS to crash.
+But will it if you run it on a CHERI system?
+
+If it's written in MISRA C and passing the checks (and not arbitrarily silencing them), it won't.
+If it's written in safe Rust, it won't.
+If it's written in SPARK Ada, it won't.
+
+Something like this will not be dynamically allocating memory, it will not be accessing arbitrary memory locations via untrusted pointers.
+It will have a set of inputs that should be amenable to exhaustive testing.
+
+If you have a chip dedicated to running the ABS and that code is written in a safe language, CHERI doesn't improve on the state of the art: you have already statically proven that nothing that CHERI dynamically prevents can happen.
+If you are building a *mixed criticality* system, the benefits of CHERI are far more pronounced: CHERI will allow you to collocate things on the same chip that previously needed to be isolated.
+
+## What does CHERIoT do?
+
+CHERIoT RTOS provides three ways for a compartment to handle CHERI failures (and other synchronous trap, such as invalid instruction errors):
+
+ - If there is enough stack space, a compartment may provide an error handler that runs with a copy of the register file and can inspect and modify the state.
+ - If there is not enough stack space or the first kind of error handler is not provided, the compartment may provide a *stackless* error handler.
+   This is used to implement our structured error handling mechanisms.
+ - If neither is provided, or the error handler returns the force-unwind result, the current compartment call is and control returns to the caller with an error code in the return register.
```

Comment:
| the current compartment call is 
what? terminated?

Reply:
Ooops.  Fixed?